Pb compute api

alibazlamit opened the pull request:
@nacx if you came across this
`INFO: << configured node(1821134a-fa51-4d3a-8a3d-da1e891dc4d2/fc621963-76bc-495f-8585-afd0885d33a9) with bash: java: command not found and jetty head: cannot open '/usr/local/jetty/VERSION.txt' for reading: No such file or directory in 27s `
`could not connect to any ip address port 8080 on node bash: line 0: cd: /usr/local/jetty: No such file or directory`

alibazlamit commented:
rebuild please

nacx commented:
I think that log is from a test that installs Jetty on a VM and then tries to connect to it, but it seems that it is failing to install Java on the node.

When jclouds configures nodes, it leaves the following files in `/tmp/bootstrap`:

* The executed script
* rc - A file with the script return code.
* stdout.log - The script output
* stderr.log - The script error output

Could you place a breakpoint in the test, so you can access the node *after* jclouds has run the scripts on it and *before* the test destroys the node? That way you should be able to retrieve those files. If you share them in a gist we'll be able to see what is going wrong in the script execution and determine the cause of the failure.

alibazlamit commented:
@nacx the stderr.log had this 
dpkg-preconfigure: unable to re-open stdin: No such file or directory
W: --force-yes is deprecated, use one of the options starting with --allow instead.
E: Package 'openjdk-6-jdk' has no installation candidate
aborting: JDK installation failed

nacx commented:
It looks like we should update the [install jdk script](https://github.com/jclouds/jclouds/blob/master/scriptbuilder/src/main/resources/functions/installOpenJDK.sh). Could you file a JIRA issue for this?

Meanwhile, could you test with an older Ubuntu version? You can force an image by running the tests with: `-Djclouds.template="osVersionMatches=15.*"` or `-Djclouds.template="imageId=an-image-id"`, to force a lookup of an Ubuntu 15.* image or to explicitly pass the id of an image.

alibazlamit commented:
Ticket here https://issues.apache.org/jira/browse/JCLOUDS-1133

alibazlamit commented:
Changing to the ubuntu 15 made it work, all the test pass now except for one `testAScriptExecutionAfterBootWithBasicTemplate `with this error message, it seems like it cannot login with the user foo, i have tried to add a wait of two minutes after server creation but no luck any thoughts?

```
(foo:rsa[fingerprint(94:4d:28:84:3a:ae:1e:39:30:f0:bb:38:44:58:44:d4),sha1(f6:ce:f4:d2:e0:bd:8a:f6:48:6c:ba:e0:4b:08:d7:c1:42:96:ac:82)]@158.222.103.139:22) (foo:rsa[fingerprint(94:4d:28:84:3a:ae:1e:39:30:f0:bb:38:44:58:44:d4),sha1(f6:ce:f4:d2:e0:bd:8a:f6:48:6c:ba:e0:4b:08:d7:c1:42:96:ac:82)]@158.222.103.139:22) error acquiring {hostAndPort=158.222.103.139:22, loginUser=foo, ssh=445447871, connectTimeout=300000, sessionTimeout=300000} (not retryable): Exhausted available authentication methods
org.jclouds.rest.AuthorizationException
	at org.jclouds.sshj.SshjSshClient.propagate(SshjSshClient.java:394)
	at org.jclouds.sshj.SshjSshClient.acquire(SshjSshClient.java:215)
	at org.jclouds.sshj.SshjSshClient.connect(SshjSshClient.java:224)
```

nacx commented:
This might be because the create node method in the adapter or the function that transforms a server into a jclouds node don't properly populate the node credentials. jclouds uses those to access the node. Could you have a look at that?

alibazlamit commented:
` loginUser=foo, ssh=445447871` its passing the credentials as you see in the log meesage.

I was able to debug the test and it fails when reaching this line 

https://github.com/jclouds/jclouds/blob/master/compute/src/test/java/org/jclouds/compute/internal/BaseComputeServiceLiveTest.java#L305

alibazlamit commented:
Full Stack trace

```
(foo:rsa[fingerprint(34:3e:d9:0d:4f:c2:78:ed:e2:60:7e:6f:1c:42:c3:88),sha1(65:81:a8:37:1c:f5:c6:70:b0:d2:d5:60:05:eb:8b:ef:53:92:d9:57)]@158.222.103.136:22) (foo:rsa[fingerprint(34:3e:d9:0d:4f:c2:78:ed:e2:60:7e:6f:1c:42:c3:88),sha1(65:81:a8:37:1c:f5:c6:70:b0:d2:d5:60:05:eb:8b:ef:53:92:d9:57)]@158.222.103.136:22) error acquiring {hostAndPort=158.222.103.136:22, loginUser=foo, ssh=1169424009, connectTimeout=300000, sessionTimeout=300000} (not retryable): Exhausted available authentication methods
org.jclouds.rest.AuthorizationException
	at org.jclouds.sshj.SshjSshClient.propagate(SshjSshClient.java:394)
	at org.jclouds.sshj.SshjSshClient.acquire(SshjSshClient.java:215)
	at org.jclouds.sshj.SshjSshClient.connect(SshjSshClient.java:224)
	at org.jclouds.compute.callables.SudoAwareInitManager.refreshAndRunAction(SudoAwareInitManager.java:74)
	at org.jclouds.compute.callables.BlockUntilInitScriptStatusIsZeroThenReturnOutput.interruptTask(BlockUntilInitScriptStatusIsZeroThenReturnOutput.java:159)
	at com.google.common.util.concurrent.AbstractFuture.cancel(AbstractFuture.java:136)
	at org.jclouds.compute.internal.BaseComputeServiceLiveTest.weCanCancelTasks(BaseComputeServiceLiveTest.java:305)
	at org.jclouds.compute.internal.BaseComputeServiceLiveTest.testAScriptExecutionAfterBootWithBasicTemplate(BaseComputeServiceLiveTest.java:262)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:497)
	at org.testng.internal.MethodInvocationHelper.invokeMethod(MethodInvocationHelper.java:85)
	at org.testng.internal.Invoker.invokeMethod(Invoker.java:696)
	at org.testng.internal.Invoker.invokeTestMethod(Invoker.java:882)
	at org.testng.internal.Invoker.invokeTestMethods(Invoker.java:1189)
	at org.testng.internal.TestMethodWorker.invokeTestMethods(TestMethodWorker.java:124)
	at org.testng.internal.TestMethodWorker.run(TestMethodWorker.java:108)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
	at java.lang.Thread.run(Thread.java:745)
Caused by: net.schmizz.sshj.userauth.UserAuthException: Exhausted available authentication methods
	at net.schmizz.sshj.SSHClient.auth(SSHClient.java:217)
	at net.schmizz.sshj.SSHClient.authPublickey(SSHClient.java:316)
	at net.schmizz.sshj.SSHClient.authPublickey(SSHClient.java:335)
	at org.jclouds.sshj.SSHClientConnection.create(SSHClientConnection.java:163)
	at org.jclouds.sshj.SSHClientConnection.create(SSHClientConnection.java:49)
	at org.jclouds.sshj.SshjSshClient.acquire(SshjSshClient.java:195)
	at org.jclouds.sshj.SshjSshClient.connect(SshjSshClient.java:224)
	at org.jclouds.compute.callables.SudoAwareInitManager.refreshAndRunAction(SudoAwareInitManager.java:74)
	at org.jclouds.compute.callables.BlockUntilInitScriptStatusIsZeroThenReturnOutput.interruptTask(BlockUntilInitScriptStatusIsZeroThenReturnOutput.java:159)
	at com.google.common.util.concurrent.AbstractFuture.cancel(AbstractFuture.java:136)
	at org.jclouds.compute.internal.BaseComputeServiceLiveTest.weCanCancelTasks(BaseComputeServiceLiveTest.java:305)
	at org.jclouds.compute.internal.BaseComputeServiceLiveTest.testAScriptExecutionAfterBootWithBasicTemplate(BaseComputeServiceLiveTest.java:262)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:497)
	at org.testng.internal.MethodInvocationHelper.invokeMethod(MethodInvocationHelper.java:85)
	at org.testng.internal.Invoker.invokeMethod(Invoker.java:696)
	at org.testng.internal.Invoker.invokeTestMethod(Invoker.java:882)
	at org.testng.internal.Invoker.invokeTestMethods(Invoker.java:1189)
	at org.testng.internal.TestMethodWorker.invokeTestMethods(TestMethodWorker.java:124)
	at org.testng.internal.TestMethodWorker.run(TestMethodWorker.java:108)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
	at java.lang.Thread.run(Thread.java:745)
Caused by: net.schmizz.sshj.userauth.UserAuthException: Timeout expired
	at net.schmizz.sshj.userauth.UserAuthException$1.chain(UserAuthException.java:33)
	at net.schmizz.sshj.userauth.UserAuthException$1.chain(UserAuthException.java:26)
	at net.schmizz.concurrent.Promise.retrieve(Promise.java:139)
	at net.schmizz.sshj.userauth.UserAuthImpl.authenticate(UserAuthImpl.java:69)
	at net.schmizz.sshj.SSHClient.auth(SSHClient.java:211)
	at net.schmizz.sshj.SSHClient.authPublickey(SSHClient.java:316)
	at net.schmizz.sshj.SSHClient.authPublickey(SSHClient.java:335)
	at org.jclouds.sshj.SSHClientConnection.create(SSHClientConnection.java:163)
	at org.jclouds.sshj.SSHClientConnection.create(SSHClientConnection.java:49)
	at org.jclouds.sshj.SshjSshClient.acquire(SshjSshClient.java:195)
	at org.jclouds.sshj.SshjSshClient.connect(SshjSshClient.java:224)
	at org.jclouds.compute.callables.SudoAwareInitManager.refreshAndRunAction(SudoAwareInitManager.java:74)
	at org.jclouds.compute.callables.BlockUntilInitScriptStatusIsZeroThenReturnOutput.interruptTask(BlockUntilInitScriptStatusIsZeroThenReturnOutput.java:159)
	at com.google.common.util.concurrent.AbstractFuture.cancel(AbstractFuture.java:136)
	at org.jclouds.compute.internal.BaseComputeServiceLiveTest.weCanCancelTasks(BaseComputeServiceLiveTest.java:305)
	at org.jclouds.compute.internal.BaseComputeServiceLiveTest.testAScriptExecutionAfterBootWithBasicTemplate(BaseComputeServiceLiveTest.java:262)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:497)
	at org.testng.internal.MethodInvocationHelper.invokeMethod(MethodInvocationHelper.java:85)
	at org.testng.internal.Invoker.invokeMethod(Invoker.java:696)
	at org.testng.internal.Invoker.invokeTestMethod(Invoker.java:882)
	at org.testng.internal.Invoker.invokeTestMethods(Invoker.java:1189)
	at org.testng.internal.TestMethodWorker.invokeTestMethods(TestMethodWorker.java:124)
	at org.testng.internal.TestMethodWorker.run(TestMethodWorker.java:108)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
	at java.lang.Thread.run(Thread.java:745)
Caused by: java.util.concurrent.TimeoutException: Timeout expired
	at net.schmizz.concurrent.Promise.retrieve(Promise.java:139)
	at net.schmizz.sshj.userauth.UserAuthImpl.authenticate(UserAuthImpl.java:69)
	at net.schmizz.sshj.SSHClient.auth(SSHClient.java:211)
	at net.schmizz.sshj.SSHClient.authPublickey(SSHClient.java:316)
	at net.schmizz.sshj.SSHClient.authPublickey(SSHClient.java:335)
	at org.jclouds.sshj.SSHClientConnection.create(SSHClientConnection.java:163)
	at org.jclouds.sshj.SSHClientConnection.create(SSHClientConnection.java:49)
	at org.jclouds.sshj.SshjSshClient.acquire(SshjSshClient.java:195)
	at org.jclouds.sshj.SshjSshClient.connect(SshjSshClient.java:224)
	at org.jclouds.compute.callables.SudoAwareInitManager.refreshAndRunAction(SudoAwareInitManager.java:74)
	at org.jclouds.compute.callables.BlockUntilInitScriptStatusIsZeroThenReturnOutput.interruptTask(BlockUntilInitScriptStatusIsZeroThenReturnOutput.java:159)
	at com.google.common.util.concurrent.AbstractFuture.cancel(AbstractFuture.java:136)
	at org.jclouds.compute.internal.BaseComputeServiceLiveTest.weCanCancelTasks(BaseComputeServiceLiveTest.java:305)
	at org.jclouds.compute.internal.BaseComputeServiceLiveTest.testAScriptExecutionAfterBootWithBasicTemplate(BaseComputeServiceLiveTest.java:262)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:497)
	at org.testng.internal.MethodInvocationHelper.invokeMethod(MethodInvocationHelper.java:85)
	at org.testng.internal.Invoker.invokeMethod(Invoker.java:696)
	at org.testng.internal.Invoker.invokeTestMethod(Invoker.java:882)
	at org.testng.internal.Invoker.invokeTestMethods(Invoker.java:1189)
	at org.testng.internal.TestMethodWorker.invokeTestMethods(TestMethodWorker.java:124)
	at org.testng.internal.TestMethodWorker.run(TestMethodWorker.java:108)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
	at java.lang.Thread.run(Thread.java:745)

```

nacx commented:
> I was able to debug the test and it fails when reaching this line

Ok, then you're hitting [JCLOUDS-1058](https://issues.apache.org/jira/browse/JCLOUDS-1058). I'll add a change to the base test class to isolate the "wait" thing in a test so it does not block the other ones. This will help having better feedback on the rest of the skipped tests.

@devcsrj At some point we'll need to have a deeper look at this.

devcsrj commented:
Hmm. Now this kinda sucks. I was under the impression before that this exception [only occurred](https://github.com/jclouds/jclouds-labs/pull/224#issuecomment-172719660) for the SOAP version of the API (that maybe the REST API did a few more stuff in the PB backend to prevent this).

alibazlamit commented:
@nacx All compute tests pass now except for the one above,could you please do a general review and check the state of this PR and give me back your feedback.
Thanks

jasminSPC commented:
@nacx is the issue above blocking this PR?

nacx commented:
@alibazlamit I've just opened https://github.com/jclouds/jclouds/pull/979 to isolate the task cancellation in a specific test, so it does not prevent other tests from running. Could you try building that branch and run the live tests again with that change, to se the *real* state of them?

alibazlamit commented:
@nacx its not preventing other tests from running, but anyways i tried to look for that branch, should build the same master branch or switch to another one?

nacx commented:
> but anyways i tried to look for that branch, should build the same master branch or switch to another one?

You can build the branch on my fork. But anyway, I plan to merge the PR tomorrow. I just need some time to run some live tests with that change (the ProfitBricks SOAP ones) and verify it works as expected.

alibazlamit commented:
@nacx All tests passed except for testWeCanCancelTasks with the same error mentioned above

jasminSPC commented:
@nacx what is the status of this PR?

nacx commented on pom.xml:
```diff
@@ -38,6 +38,7 @@
 
   <properties>
     <jclouds.version>2.0.0-SNAPSHOT</jclouds.version>
+	<checkstyle.skip>true</checkstyle.skip>
```
This needs to be removed

nacx commented on profitbricks-rest/pom.xml:
```diff
@@ -36,6 +36,7 @@
         <test.profitbricks-rest.identity>FIXME</test.profitbricks-rest.identity>
         <test.profitbricks-rest.credential>FIXME</test.profitbricks-rest.credential>
         <test.profitbricks-rest.api-version>1.3</test.profitbricks-rest.api-version>
+        <checkstyle.skip>true</checkstyle.skip>
```
This needs to be removed

nacx commented on profitbricks-rest/pom.xml:
```diff
@@ -108,6 +121,13 @@
+        <dependency>
+            <groupId>org.apache.jclouds.driver</groupId>
+            <artifactId>jclouds-log4j</artifactId>
+            <version>2.0.0-SNAPSHOT</version>
```
Use `${jclouds.version}`

nacx commented on profitbricks-rest/src/main/java/org/apache/jclouds/profitbricks/rest/compute/ProfitBricksComputeServiceAdapter.java:
```diff
@@ -0,0 +1,563 @@
+                @Override
+                public boolean apply(Lan input) {
+                    input = api.lanApi().get(dataCenterId, input.id(), new DepthOptions().depth(3));
+                    boolean match = input.properties().isPublic();
```
Just `return input.properties().isPublic();`

nacx commented on profitbricks-rest/src/main/java/org/apache/jclouds/profitbricks/rest/compute/ProfitBricksComputeServiceAdapter.java:
```diff
@@ -0,0 +1,563 @@
+
+        //add a NIC to the server
+        waitDcUntilAvailable.apply(dataCenterId);
+        waitServerUntilAvailable.apply(ServerRef.create(dataCenterId, serverId));
```
Is this needed? The server was not touched since the last wait operation.

nacx commented on profitbricks-rest/src/main/java/org/apache/jclouds/profitbricks/rest/compute/ProfitBricksComputeServiceAdapter.java:
```diff
@@ -0,0 +1,563 @@
+        }
+
+        waitDcUntilAvailable.apply(dataCenterId);
+        waitServerUntilAvailable.apply(ServerRef.create(dataCenterId, serverId));
```
No need to wait again here

nacx commented on profitbricks-rest/src/main/java/org/apache/jclouds/profitbricks/rest/compute/ProfitBricksComputeServiceAdapter.java:
```diff
@@ -0,0 +1,563 @@
+                // delete unconnected volume
+                logger.warn(ex, ">> failed to connect volume '%s'. deleting..", volumeId);
+                destroyVolume(volumeId, dataCenterId);
+            }
```
Shouldn't the entire process be rolled-back if we fail to create a volume?

devcsrj replied:
Prior this volume-creating code block, a [*bootVolume*](https://github.com/jclouds/jclouds-labs/pull/292/files#diff-3cfe9775f2121ca815ae0fce43f9d211R187) is attached to the server. That server-volume pair already constitutes a jcloud node. The additional volumes here are more like, additional storages, not necessarily making the node unusable if they fail in attaching.

nacx commented on profitbricks-rest/src/main/java/org/apache/jclouds/profitbricks/rest/compute/ProfitBricksComputeServiceAdapter.java:
```diff
@@ -0,0 +1,563 @@
+        ServerInDataCenter server = getNode(serverInDataCenterId);
+        // Last paranoid check
+        waitDcUntilAvailable.apply(dataCenterId);
+        waitServerUntilAvailable.apply(ServerRef.create(dataCenterId, serverId));
```
Really needed? They should be already available since last check.

nacx commented on profitbricks-rest/src/main/java/org/apache/jclouds/profitbricks/rest/compute/ProfitBricksComputeServiceAdapter.java:
```diff
@@ -0,0 +1,563 @@
+    }
+
+    @Override
+    public Iterable<DataCenter> listLocations() {
```
Locations are configured via properties, so this method should never be invoked. Let this method just throw an Unsupported op exception.

alibazlamit replied:
@nacx  Can you give me an example on how to configure locations via properties?
Thanks.

nacx replied:
Sure. Have a look at https://github.com/jclouds/jclouds/commit/7202e0557f620c272feaca3e9a49b6e45a77a791 . That commit configures the compute service to get the locations from the properties defined in the provider metadata. The important bits there are:

* The jclouds location properties configured in the provider metadata.
* The type for the locations in the generics of the adapter changes from DataCenter to the jclouds Location.
* The `LocationsFromComputeServiceAdapterModule` must be removed from the compute module. This will make the default jclouds location suppliers active and return the locations from the configured properties. You no longer need the listLocations method, so you can safely make it throw an exception.
* Since ProfitBricks needs a Datacenter, that commit declares a custom `CreateNodesInGroupThenAddToSet` strategy to take care of creating the datacenter if none exists in the selected location *before* creating any node. You can have a look at the binding and copy most of the code in that strategy (if not all).
* To let the datacenter be created *before* the adapter is called, that commit introduces an internal class that adds the created DataCenter to the jclouds Template, so it can be directly passed to the adapter, removing the need to perform additional calls.

Feel free to ask if you have any doubts or need some more guidance!

alibazlamit replied:
@nacx do i need to change the ProfitBricksComputeServiceLiveTest.java to like the one in the commit above?

nacx replied:
Yes :)

nacx commented on profitbricks-rest/src/main/java/org/apache/jclouds/profitbricks/rest/compute/config/ProfitBricksComputeServiceContextModule.java:
```diff
@@ -0,0 +1,335 @@
+        super.configure();
+
+        install(new LocationsFromComputeServiceAdapterModule<ServerInDataCenter, Hardware, Provisionable, DataCenter>() {
+        });
```
Remove this to let the locations be configured from the properties int he metadata.

nacx commented on profitbricks-rest/src/main/java/org/apache/jclouds/profitbricks/rest/compute/config/ProfitBricksComputeServiceContextModule.java:
```diff
@@ -0,0 +1,335 @@
+        }).to(ProfitBricksComputeServiceAdapter.class);
+
+        bind(new TypeLiteral<Function<org.apache.jclouds.profitbricks.rest.domain.Location, Location>>() {
+        }).to(LocationToLocation.class);
```
This can be removed too.

nacx commented on profitbricks-rest/src/main/java/org/apache/jclouds/profitbricks/rest/compute/config/ProfitBricksComputeServiceContextModule.java:
```diff
@@ -0,0 +1,335 @@
+        bind(new TypeLiteral<Function<ServerInDataCenter, NodeMetadata>>() {
+        }).to(ServerInDataCenterToNodeMetadata.class);
+        bind(new TypeLiteral<Function<Server, NodeMetadata>>() {
+        }).to(ServerToNodeMetadata.class);
```
Is this function still used? If not, remove the binding and the function class.

nacx commented on profitbricks-rest/src/main/java/org/apache/jclouds/profitbricks/rest/compute/function/DataCenterToLocation.java:
```diff
@@ -0,0 +1,52 @@
+import org.jclouds.domain.LocationBuilder;
+import org.jclouds.domain.LocationScope;
+
+public class DataCenterToLocation implements Function<DataCenter, Location> {
```
Locations are configured by properties. This class should no longer be needed and could be removed.

nacx commented on profitbricks-rest/src/main/java/org/apache/jclouds/profitbricks/rest/compute/function/LocationToLocation.java:
```diff
@@ -0,0 +1,46 @@
+import org.jclouds.domain.LocationScope;
+import org.jclouds.location.suppliers.all.JustProvider;
+
+public class LocationToLocation implements Function<Location, org.jclouds.domain.Location> {
```
Same as before. Can this class be removed?

nacx commented on profitbricks-rest/src/main/java/org/apache/jclouds/profitbricks/rest/compute/function/ServerToNodeMetadata.java:
```diff
@@ -0,0 +1,169 @@
+import org.jclouds.util.InetAddresses2.IsPrivateIPAddress;
+import static com.google.common.base.Preconditions.checkNotNull;
+
+public class ServerToNodeMetadata implements Function<Server, NodeMetadata> {
```
If this function is not used (since the adapter uses the `ServerInDatacenter` class), remove it

nacx commented on profitbricks-rest/src/test/java/org/apache/jclouds/profitbricks/rest/compute/ProfitBricksTemplateBuilderLiveTest.java:
```diff
@@ -0,0 +1,106 @@
+        System.setProperty("http.proxyPort", "8888");
+        System.setProperty("https.proxyPort", "8888");
+        System.setProperty("javax.net.ssl.trustStore", "C:\\Program Files\\Java\\jdk1.8.0_73\\jre\\lib\\security\\FiddlerKeystore");
+        System.setProperty("javax.net.ssl.trustStorePassword", "testme");
```
Remove all these properties.

nacx commented:
Apologies for the delay @alibazlamit @jasminSPC, I've been on vacation.

@devcsrj Could you help reviewing this? I can mostly review syntax, etc, but you have a better knowledge of how the provider works for a proper functional review.

alibazlamit commented:
@nacx should i be following [this ](https://github.com/jclouds/jclouds/commit/7202e0557f620c272feaca3e9a49b6e45a77a791) apply those changes ?

nacx commented:
> should i be following this apply those changes ?

Yes. I just added a comment before having read this one :) I've tried to summarise the intent of the important bits in that commit, but feel free to ask for help/clarification, etc!

alibazlamit commented:
rebuild please

alibazlamit commented:
@nacx i have applied the changes above please take a look.
Thanks.

devcsrj commented:
Hi @alibazlamit !

The initial comments on this PR haven't been addressed yet; see [Files Changed](https://github.com/jclouds/jclouds-labs/pull/292/files) tab of this PR, most notably style-checks. The diff made it looked like that this PR changed *everything* due to the formatting changes. While this does not necessarily break code correctness, reviewing becomes 10x more times difficult (actually, 5700+ additions ;) ), as we're *might* be seeing changes not intentionally addressed by this PR (i.e.: Compute Service).

Could you take care of that first, by removing the disabled `<checkstyle.skip>`s, and reformatting the source? If you're using eclipse or idea, they have a profile you can format [here](https://cwiki.apache.org/confluence/display/JCLOUDS/Coding+Standards).

devcsrj commented on profitbricks-rest/src/main/java/org/apache/jclouds/profitbricks/rest/compute/ProfitBricksComputeServiceAdapter.java:
```diff
@@ -0,0 +1,568 @@
+        TemplateOptions options = template.getOptions();
+        final String loginUser = isNullOrEmpty(options.getLoginUser()) ? "root" : options.getLoginUser();
+//        final String password = options.hasLoginPassword() ? options.getLoginPassword() : Passwords.generate();
+        final String password = "LTbAHNbcMt";
```
Is... this password suppose to be here?

devcsrj commented on profitbricks-rest/src/main/java/org/apache/jclouds/profitbricks/rest/compute/ProfitBricksComputeServiceAdapter.java:
```diff
@@ -0,0 +1,568 @@
+                final org.apache.jclouds.profitbricks.rest.domain.Volume.Request.CreatePayload request = org.apache.jclouds.profitbricks.rest.domain.Volume.Request.creatingBuilder()
+                        .dataCenterId(dataCenterId)
+                        // put image to first volume
+                        .image("02a49ebb-27b7-11e6-8e88-52540005ab80")
```
Take the image id from the template options instead

devcsrj commented on profitbricks-rest/src/main/java/org/apache/jclouds/profitbricks/rest/compute/ProfitBricksComputeServiceAdapter.java:
```diff
@@ -0,0 +1,568 @@
+        try {
+            Thread.sleep(90000);
+        } catch (InterruptedException ex) {
+            java.util.logging.Logger.getLogger(ProfitBricksComputeServiceAdapter.class.getName()).log(Level.SEVERE, null, ex);
```
Use the injected `logger` field

devcsrj commented on profitbricks-rest/src/main/java/org/apache/jclouds/profitbricks/rest/compute/ProfitBricksComputeServiceAdapter.java:
```diff
@@ -0,0 +1,568 @@
+        waitServerUntilAvailable.apply(ServerRef.create(dataCenterId, serverId));
+
+        try {
+            Thread.sleep(90000);
```
Also, why the need for 90-second sleep here? If the provisioning process still not done here (*sigh*), adjust instead the default values of the `waitServerUntilAvailable` predicate or datacenter, whichever applicable.

devcsrj commented on profitbricks-rest/src/test/java/org/apache/jclouds/profitbricks/rest/compute/BaseTest.java:
```diff
@@ -0,0 +1,948 @@
+import org.testng.annotations.Test;
+
+@Test(groups = {"integration", "live"}, singleThreaded = true)
+public abstract class BaseTest extends BaseComputeServiceContextLiveTest {
```
The ssh-related issue was already fixed by @nacx with [https://github.com/jclouds/jclouds/pull/995], so you can revert the the overridden methods in this test class.

devcsrj commented on profitbricks-rest/src/test/java/org/apache/jclouds/profitbricks/rest/compute/ProfitBricksComputeServiceLiveTest.java:
```diff
@@ -0,0 +1,156 @@
+//        System.setProperty("http.proxyPort", "8888");
+//        System.setProperty("https.proxyPort", "8888");
+//        System.setProperty("javax.net.ssl.trustStore", "C:\\Program Files\\Java\\jdk1.8.0_73\\jre\\lib\\security\\FiddlerKeystore");
+//        System.setProperty("javax.net.ssl.trustStorePassword", "testme");
```
Remove these commented test values (including the `TEST_DC_NAME` field)

devcsrj commented on profitbricks-rest/src/test/java/org/apache/jclouds/profitbricks/rest/compute/ProfitBricksComputeServiceLiveTest.java:
```diff
@@ -0,0 +1,156 @@
+//        templateBuilder.osVersionMatches("16.04");
+//        templateBuilder.locationId(dataCenter.id());
+//        templateBuilder.hardwareId("cpu=4,ram=2,disk=30");
+//        templateBuilder.imageId("02a49ebb-27b7-11e6-8e88-52540005ab80");
```
Remove these as well

devcsrj commented on profitbricks-rest/src/main/java/org/apache/jclouds/profitbricks/rest/compute/config/ProfitBricksComputeServiceContextModule.java:
```diff
@@ -0,0 +1,333 @@
+        ComputeServiceAdapterContextModule<ServerInDataCenter, Hardware, Provisionable, DataCenter> {
+
+    @Override
+    protected void configure() {
```
Now that jclouds supports arbitrary CPU, RAM and Storage, configure the template builder here too! See [blog post](https://jclouds.apache.org/blog/2016/08/22/arbitrary-cpu-ram/) for details.

alibazlamit replied:
@devcsrj  can i use [this ](https://github.com/jclouds/jclouds/tree/master/providers/profitbricks)as a reference?

alibazlamit commented:
@devcsrj the support for  arbitrary CPU, RAM and Storage was done something was not right with the branch i pushed, now i have updated it to the upstream and squashed all the changes in one commit, one issue i noticed is that its diffing white spaces and eol, i have configure eol to be lf but still its showing me a lot of changes maybe @nacx can help too on this one.
Other than the diff issue the branch is ready for reviewing.

Thanks

devcsrj commented:
Now that's a **LOT** of formatting noise, even GitHub couldn't render. :(

![](https://cloud.githubusercontent.com/assets/3963900/18572062/51bd9a1c-7bea-11e6-9d97-1a92f8b94a9f.png)

Assuming you used the IDE formatting profiles, then the formatting noise is most probably caused by earlier PB-related PRs not formatted properly. i.e.: Braces on single-line ifs:

![](https://cloud.githubusercontent.com/assets/3963900/18572146/205b894c-7beb-11e6-9db5-6e85acea279a.png)

..or untrimmed whitespaces:
![](https://cloud.githubusercontent.com/assets/3963900/18572159/47b3cf22-7beb-11e6-87dd-e5aaf9ac81b6.png)

Here's what I suggest, to move this PR into a reviewable state:
- Rebase this branch to `master`
- Re-apply **ONLY** the compute-api related changes, using the correct formatting

Don't reformat the rest of the codebase _yet_; only the once you've changed. After this gets merged, we'll dedicated a separate PR for the sole purpose of reformatting the rest of the codebase. That way we keep the changes to it's own PRs, making the review possible.

alibazlamit commented:
@devcsrj I removed all the formatting changes, i agree lets focus on the compute changes, all the changes in the commit are related to the Compute API .

devcsrj commented on profitbricks-rest/src/main/java/org/apache/jclouds/profitbricks/rest/compute/ProfitBricksComputeServiceAdapter.java:
```diff
@@ -0,0 +1,570 @@
+            logger.trace("<< provisioning volume '%s'", volume);
+            final org.apache.jclouds.profitbricks.rest.domain.Volume.Request.CreatePayload request = org.apache.jclouds.profitbricks.rest.domain.Volume.Request.creatingBuilder()
+                    .dataCenterId(dataCenterId)
+                    .image(image.getId())
```
Is it intentional that here we provision every volume with an image? (i.e.: If we attached 5 volumes, each has a bootable OS?)

devcsrj commented on profitbricks-rest/src/main/java/org/apache/jclouds/profitbricks/rest/compute/ProfitBricksComputeServiceAdapter.java:
```diff
@@ -0,0 +1,570 @@
+      api.serverApi().attachVolume(Server.Request.attachVolumeBuilder()
+              .dataCenterId(dataCenterId)
+              .serverId(serverId)
+              .volumeId(bootVolume.id())
```
IIRC, if one creates a server (line 193) with the `bootVolume` param set, it'll be automatically attached after that server is provisioned. Is this not the case anymore?

alibazlamit replied:
Sorry the comment below was supposed to be here :)

> It is true that's the case, for some reason doing that cause the live tests to fail, more specifically when trying to ssh into the servers and test operations, as if the server was not ready yet, i tried using long sleep's and checks nothing helped except for this approach, it does the job don't you agree?

devcsrj replied:
> when trying to ssh into the servers and test operations..

Ohhhhhhhh. I guess we'll leave it at that! :)

devcsrj commented on profitbricks-rest/src/main/java/org/apache/jclouds/profitbricks/rest/compute/ProfitBricksComputeServiceAdapter.java:
```diff
@@ -0,0 +1,570 @@
+
+   @Override
+   public Provisionable getImage(String id
+   ) {
```
Woops right paren

alibazlamit replied:
It is true that's the case, for some reason doing that cause the live tests to fail, more specifically when trying to ssh into the servers and test operations, as if the server was not ready yet, i tried using long sleep's and checks nothing helped except for this approach, it does the job don't you agree?

devcsrj replied:
> ..for this approach

Uhm.. Which approach again? (Wrong comment thread?)

devcsrj commented on profitbricks-rest/src/main/java/org/apache/jclouds/profitbricks/rest/domain/Location.java:
```diff
@@ -17,12 +17,13 @@
    US_LASDEV("us/lasdev", "USA Developer cluster"),
-   UNRECOGNIZED("unrecognized", "Unrecognized location");
+   UNRECOGNIZED("unrecognized", "Unrecognized location"),
+   Mock("mock", "mock");
```
[nit] All caps

devcsrj commented on profitbricks-rest/src/main/java/org/apache/jclouds/profitbricks/rest/domain/Location.java:
```diff
@@ -49,9 +46,16 @@ public static Location fromValue(String v) {
+
+   @Override
+   public String toString() {
+      return id;
```
Make this more descriptive, something like:
```
   return "Location[id=" + id + ", description=" + description + "]";
```

..or don't override it at all, and use the actual `getId()` method, to print out the ID.

devcsrj commented on profitbricks-rest/pom.xml:
```diff
@@ -61,9 +66,9 @@
-           <version>${jclouds.version}</version>
+            <groupId>org.apache.jclouds.driver</groupId>
+            <artifactId>jclouds-okhttp</artifactId>
+            <version>${jclouds.version}</version>
```
[nit] Formatting

alibazlamit commented:
@devcsrj the gist results [here](https://gist.github.com/alibazlamit/c09d8fc26af8d615bf7f48c710a65254)

devcsrj commented on profitbricks-rest/src/main/java/org/apache/jclouds/profitbricks/rest/ProfitBricksProviderMetadata.java:
```diff
@@ -55,41 +55,44 @@ public ProfitBricksProviderMetadata(Builder builder) {
+      properties.setProperty(PROPERTY_ZONE + ".de/fkb." + ISO3166_CODES, "DE-BW");
+      properties.setProperty(PROPERTY_ZONE + ".de/fra." + ISO3166_CODES, "DE-HE");
+      properties.setProperty(PROPERTY_ZONE + ".us/las." + ISO3166_CODES, "US-NV");
+      properties.setProperty(PROPERTY_ZONE + ".us/lasdebv." + ISO3166_CODES, "US-NV");
```
`.us/lasdebv` is this intentional?

alibazlamit replied:
Mistake it should be `.us/lasdev`

alibazlamit commented:
@devcsrj How close would you say we are to closing this PR?

alibazlamit commented:
rebuild please

alibazlamit commented:
rebuild please

alibazlamit commented:
please rebuild

baldwinSPC commented:
@devcsrj *nudge* ;)

devcsrj commented:
Hello! 

> How close would you say we are to closing this PR

I'd say it's long overdue! Seeing your live tests passing from your gist, I'd say the impl addresses the contract required by the `ComputeService`. Just some last nitpicks like that [overriden toString](https://github.com/jclouds/jclouds-labs/pull/292#discussion_r79763475), and the [formatting in pom](https://github.com/jclouds/jclouds-labs/pull/292#discussion_r79763965). Otherwise, I'd say this LGTM.

/cc @nacx PR now yours for last review?

---
What're your plans btw after this gets merged? I'd personally like to see:
- Arbitrary CPU/RAM support, as this is [now supported](https://jclouds.apache.org/blog/2016/08/22/arbitrary-cpu-ram/)
- [JCLOUDS-1082](https://issues.apache.org/jira/browse/JCLOUDS-1082) Firewall configurations
- [JCLOUDS-1083](https://issues.apache.org/jira/browse/JCLOUDS-1083) Inbound port configurations

alibazlamit commented:
@devcsrj this is the gist including the arbitrary support tests [here ](https://gist.github.com/alibazlamit/0dcbdcf91009c0fc4c54b2878ed6b512), all the nitpicks have been changed and pushed, i hope we will be able to close this one soon as you said its long overdue.

Thanks.

nacx commented on profitbricks-rest/src/main/java/org/apache/jclouds/profitbricks/rest/compute/ProfitBricksComputeServiceAdapter.java:
```diff
@@ -0,0 +1,577 @@
+               destroyServer(serverId, dataCenterId);
+               throw ex;
+            } catch (Exception ex1) {
+               java.util.logging.Logger.getLogger(ProfitBricksComputeServiceAdapter.class.getName()).log(Level.SEVERE, null, ex1);
```
Change to the jclouds logger used in this class.

alibazlamit replied:
Done, changed.

alibazlamit replied:
Updated to the latest just let me know if you need me to squash this, Thanks for the review.

alibazlamit commented:
Updated to the latest just let me know if you need me to squash this, Thanks for the review.

nacx commented:
Squashed and pushed to master as [5742745e](http://git-wip-us.apache.org/repos/asf/jclouds-labs/commit/5742745e). Thanks @alibazlamit @devcsrj!
